pep-9999.rst

```diff
@@ -0,0 +1,341 @@
+-------------------
+
+Import side effects that would otherwise be produced by the execution of
+imported modules during the execution of import statements will be deferred at
+least until the imported objects are used.
+
+These import side effects may include:
+
+* code executing any side-effecting logic during import;
+* relying on imported submodules being set as attributes in the parent module.
+
+
+Dynamic Paths
+-------------
+
+There could be issues related to dynamic Python import paths; particularly,
+adding (and then removing after the import) paths from ``sys.path``.
+
+.. code-block py
+    sys.path.insert(0, "/path/to/foo/module")
+    import foo
+    del sys.path[0]
+    foo.Bar()
+
+In this case, with lazy imports enabled, the import of ``foo`` will not
+actually occur while the addition to ``sys.path`` is present.
+
+
+Deferred Exceptions
+-------------------
+
+All exceptions arising from import (including ``ModuleNotFoundError``) are
+deferred from import time to first-use time, which might complicate debugging.
+Accessing an object in the middle of any code could trigger a deferred import
+and produce ``ImportError`` or any other exception resulting from the
+resolution of the deferred object, while loading and executing the related
+imported module.
+
+
+Security Implications
+=====================
+
+Deferred execution of code could produce security concerns if process owner,
+path, ``sys.path``, or other sensitive environment or contextual states change
+between the time the ``import`` statement is executed and the time where the
+imported object is used.
+
+
+Performance Impact
+==================
+
+The reference implementation has shown that the feature has negligible
+performance impact on existing real-world codebases (Instagram Server and other
+several CLI programs at Meta), while providing substantial improvements to
+startup time and memory usage.
+
+The reference implementation shows small performance regressions in a few
+pyperformance benchmarks, but improvements in others. (TODO update with
+detailed data from 3.11 port of implementation.)
+
+
+How to Teach This
+=================
+
+In most cases, lazy imports should just work transparently and no teaching of
+the feature should be necessary.
+
+The implementation will ensure that errors resulting from a deferred import
+have metadata attached pointing the user to the original import statement, to
+ease debuggability of errors from lazy imports.
+
+Some best practices to deal with some of the issues that could arise and to
+better take advantage of lazy imports are:
+
+* Avoid relying on import side effects. Perhaps the most common reliance on
+  import side effects is the registry pattern, where population of some
+  external registry happens implicitly during the importing of modules, often
+  via decorators. Instead, the registry should be built via an explicit call
+  that perhaps does a discovery process to find decorated functions or classes.
```

Comment:
Does this break the `click` use case?  The PEP should at least discuss this topic, as CLIs are the primary motivator and `click` is a very common CLI library.

Reply:
Yes, if click subcommands are created in multiple modules, those modules need to be forced to eager import in order to get the click subcommands registered. Will note this specifically.